File: src/eval/compression.ts
import type { CompressionExample } from "./examples.js";

export interface CompressionStats {
  naturalLanguageChars: number;
  glyphChars: number;
  naturalLanguageWords: number;
  glyphWords: number;
  naturalLanguageApproxTokens: number;
  glyphApproxTokens: number;
  compressionRatio: number;
}

export function compareCompression(glyphSource: string, example: CompressionExample): CompressionStats {
  const naturalLanguage = example.naturalLanguage.trim();
  const glyph = glyphSource.trim();

  return {
    naturalLanguageChars: naturalLanguage.length,
    glyphChars: glyph.length,
    naturalLanguageWords: countWords(naturalLanguage),
    glyphWords: countWords(glyph),
    naturalLanguageApproxTokens: approximateTokens(naturalLanguage),
    glyphApproxTokens: approximateTokens(glyph),
    compressionRatio: naturalLanguage.length / Math.max(1, glyph.length)
  };
}

export function countWords(text: string): number {
  const words = text.trim().match(/\S+/g);
  return words ? words.length : 0;
}

export function approximateTokens(text: string): number {
  return Math.ceil(text.length / 4);
}

File: src/eval/compression.rs
use serde::Serialize;

use super::examples::CompressionExample;

#[derive(Debug, Clone, PartialEq, Serialize)]
pub struct CompressionStats {
    #[serde(rename = "naturalLanguageChars")]
    pub natural_language_chars: usize,
    #[serde(rename = "glyphChars")]
    pub glyph_chars: usize,
    #[serde(rename = "naturalLanguageWords")]
    pub natural_language_words: usize,
    #[serde(rename = "glyphWords")]
    pub glyph_words: usize,
    #[serde(rename = "naturalLanguageApproxTokens")]
    pub natural_language_approx_tokens: usize,
    #[serde(rename = "glyphApproxTokens")]
    pub glyph_approx_tokens: usize,
    #[serde(rename = "compressionRatio")]
    pub compression_ratio: f64,
}

pub fn compare_compression(glyph_source: &str, example: &CompressionExample) -> CompressionStats {
    let natural_language = example.natural_language.trim();
    let glyph = glyph_source.trim();

    CompressionStats {
        natural_language_chars: natural_language.len(),
        glyph_chars: glyph.len(),
        natural_language_words: count_words(natural_language),
        glyph_words: count_words(glyph),
        natural_language_approx_tokens: approximate_tokens(natural_language),
        glyph_approx_tokens: approximate_tokens(glyph),
        compression_ratio: natural_language.len() as f64 / glyph.len().max(1) as f64,
    }
}

pub fn count_words(text: &str) -> usize {
    text.split_whitespace().count()
}

pub fn approximate_tokens(text: &str) -> usize {
    text.len().div_ceil(4)
}
